fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,809 @@
+fn update_wait_on_escape_ms(vars: &dyn Environment) {
+    let fish_escape_delay_ms = vars.get_unless_empty(L!("fish_escape_delay_ms"));
+    let var = crate::env::environment::env_var_to_ffi(fish_escape_delay_ms);
+    crate::ffi::update_wait_on_escape_ms_ffi(var);
+}
+
+impl VarDispatchTable {
+    fn observes_var(&self, name: &wstr) -> bool {
+        self.named_table.contains_key(name) || self.anon_table.contains_key(name)
+    }
+
+    /// Add a callback for the variable `name`. We must not already be observing this variable.
+    pub fn add(&mut self, name: &'static wstr, callback: NamedEnvCallback) {
+        let prev = self.named_table.insert(name, callback);
+        assert!(prev.is_none(), "Already observing {}", name);
+    }
+
+    /// Add an callback for the variable `name`. We must not already be observing this variable.
+    pub fn add_anon(&mut self, name: &'static wstr, callback: AnonEnvCallback) {
+        let prev = self.anon_table.insert(name, callback);
+        assert!(prev.is_none(), "Already observing {}", name);
+    }
+
+    pub fn dispatch(&self, key: &wstr, vars: &EnvStack) {
+        if let Some(named) = self.named_table.get(key) {
+            (named)(key, vars);
+        }
+        if let Some(anon) = self.anon_table.get(key) {
+            (anon)(vars);
+        }
+    }
+}
+
+fn handle_timezone(var_name: &wstr, vars: &dyn Environment) {
+    let var = vars.get_unless_empty(var_name).map(|v| v.as_string());
+    FLOGF!(
+        env_dispatch,
+        "handle_timezone() current timezone var:",
+        var_name,
+        "=>",
+        var.as_ref()
+            .map(|v| v.as_utfstr())
+            .unwrap_or(L!("MISSING/EMPTY")),
+    );
+    let name = var_name.to_string();
+    if let Some(value) = var {
+        let value = crate::common::wcs2zstring(&value);
+        setenv_lock(&name, &value, true);
+    } else {
+        unsetenv_lock(&name);
+    }
+
+    extern "C" {
+        fn tzset();
+    }
+
+    unsafe {
+        tzset();
+    }
+}
+
+/// Update the value of [`FISH_EMOJI_WIDTH`].
+fn guess_emoji_width(vars: &dyn Environment) {
+    use crate::fallback::FISH_EMOJI_WIDTH;
+
+    if let Some(width_str) = vars.get(L!("fish_emoji_width")) {
+        // The only valid values are 1 or 2; we default to 1 if it was an invalid int.
+        let new_width = fish_wcstoi(width_str.as_string().chars()).unwrap_or(1);
+        let new_width = new_width.clamp(1, 2);
+        FISH_EMOJI_WIDTH.store(new_width, Ordering::Relaxed);
+        FLOGF!(
+            term_support,
+            "Overriding default fish_emoji_width w/",
+            new_width
+        );
+        return;
+    }
+
+    let term = vars
+        .get(L!("TERM_PROGRAM"))
+        .map(|v| v.as_string())
+        .unwrap_or_else(WString::new);
+    let version = vars
+        .get(L!("TERM_PROGRAM_VERSION"))
+        .map(|v| v.as_string().to_string())
+        .and_then(|v| v.parse().ok())
```

Comment:
This was `strtod` in the previous code and should be `wcstod` in Rust.

For example my `$TERM_PROGRAM_VERSION` is `3.4.19`; this will fail with `parse()` but succeeds with `wcstod`.

Reply:
I was confused by the existing code and diverged from the C++ codebase. I should have mentioned the change. 

I ended up looking at what we were using `TERM_PROGRAM_VERSION` for and what the valid contents of that string are for the cases we are interested in and found that for `TERM_PROGRAM == "Apple_Terminal"` (which is the only case the code cares about) the version would be an integral value (we check for `version >= 400`).

tbh I have been trying to move away in the little places that I can from our custom code where the standard library functions will suffice.

Reply:
Moving towards standard library functions may be a worthy goal, but it should be done as a separately factored commit, not in this one.